mlflow/deployments/databricks/__init__.py

```diff
@@ -1,39 +1,123 @@
 
 
 class DatabricksDeploymentClient(BaseDeploymentClient):
     def create_deployment(self, name, model_uri, flavor=None, config=None, endpoint=None):
+        """
+        .. warning::
+
+            This method is not implemented for Databricks deployments.
+        """
         raise NotImplementedError
 
     def update_deployment(self, name, model_uri=None, flavor=None, config=None, endpoint=None):
+        """
+        .. warning::
+
+            This method is not implemented for Databricks deployments.
+        """
         raise NotImplementedError
 
     def delete_deployment(self, name, config=None, endpoint=None):
+        """
+        .. warning::
+
+            This method is not implemented for Databricks deployments.
+        """
         raise NotImplementedError
 
     def list_deployments(self, endpoint=None):
+        """
+        .. warning::
+
+            This method is not implemented for Databricks deployments.
+        """
         raise NotImplementedError
 
     def get_deployment(self, name, endpoint=None):
+        """
+        .. warning::
+
+            This method is not implemented for Databricks deployments.
+        """
         raise NotImplementedError
 
+    def _call_endpoint(
+        self,
+        *,
+        method: str,
+        prefix: str = "/api/2.0",
+        route: Optional[str] = None,
+        json_body: Optional[Dict[str, Any]] = None,
+    ):
+        call_kwargs = {}
+        if method.lower() == "get":
+            call_kwargs["params"] = json_body
+        else:
+            call_kwargs["json"] = json_body
+
+        response = http_request(
+            host_creds=get_databricks_host_creds("databricks"),
+            endpoint=posixpath.join(prefix, "serving-endpoints", route if route else ""),
+            method=method,
+            timeout=10,
+            raise_on_status=False,
+            **call_kwargs,
+        )
+        augmented_raise_for_status(response)
+        return response.json()
+
     def predict(self, deployment_name=None, inputs=None, endpoint=None):
-        raise NotImplementedError("TODO")
+        """
+        TODO
+        """
+        return self._call_endpoint(
+            method="POST",
+            prefix="/",
+            route=posixpath.join(endpoint, "invocations"),
+            json_body=inputs,
+        )
 
     def create_endpoint(self, name, config=None):
-        raise NotImplementedError("TODO")
+        """
+        TODO
+        """
+        config = config.copy() if config else {}
+        payload = {"name": name, "config": config}
+        if task := config.pop("task", None):
+            payload["task"] = task
+        return self._call_endpoint(method="POST", json_body=payload)
 
     def update_endpoint(self, endpoint, config=None):
-        raise NotImplementedError("TODO")
+        """
+        TODO
+        """
+        return self._call_endpoint(
+            method="PUT", route=posixpath.join(endpoint, "config"), json_body=config
```

Comment:
@sueann To update rate limits, are users expected to supply the whole config?

Reply:
yeah. we could have made it a PATCH update but i think the decision to make it a PUT was because the `key` is not well-defined (??). @mingyu89?

Reply:
Yes. Here's example of the payload data
```
data = {
    "rate_limits": [{"key": "user", "renewal_period": "minute", "calls": 5}]
}
```
If we do patch, we have to compare all three fields to do the remove/update.
And currently the max number of limits people can specify is 2 - one user limit and one endpoint limit.

Reply:
Rate limit PUT has its own request.
<img width="1219" alt="Screenshot 2023-11-16 at 12 31 49 PM" src="https://github.com/mlflow/mlflow/assets/12734110/6d896ee0-b581-4e85-a970-c233397f5b7e">

Reply:
@mingyu89 Thanks! Is this a new endpoint we added for setting rate limits?

Reply:
It's a new API for updating(PUT) rate limits. 
Rate limits can also be specified during endpoint creation.
The rate limits can only be set on foundation models and external models.

Reply:
I updated the notebook to include creating endpoint with rate limits and set rate limits.
Setting rate limit with empty array is to delete the rate limits. https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/4169768892691893/command/4169768892691894

Reply:
@mingyu89 Is this a valid request?

```
url = (
    host + "api/2.0/serving-endpoints/mingyu_external_openai_embeddings_endpoint/config"
)

data = {
    "name": "mingyu_openai_embeddings_endpoint",
    "served_entities": [
        {
            "name": "updated-served-entity-name",
            "external_model": {
                "name": "text-embedding-ada-002",
                "provider": "openai",
                "task": "llm/v1/embeddings",
                "openai_config": {
                    "openai_api_key": "{{secrets/mingyu-external-model/openai}}"
                },
            },
        }
    ],
    "rate_limits": [
        {
            "key": "user",
            "renewal_period": "minute",
            "calls": 3,
        }
    ],
}

response = requests.put(url, headers=headers, json=data)
```

I got this error:

```
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://e2-dogfood.staging.cloud.databricks.com/api/2.0/serving-endpoints/test-endpoint-816e1f87-2deb-4e09-a69e-0656703932f9/config. Response text: {"error_code":"BAD_REQUEST","message":"Config needs to be set and config.served_models must contain at least one element"}
```

Reply:
Never mind. I found a mistake.